src/bidiMapper/BidiServer.ts

```diff
@@ -22,73 +22,72 @@ import {ITransport} from '../utils/transport';
   message: Message.RawCommandRequest;
 };
 
-export class BiDiMessageEntry {
-  readonly #message: Message.OutgoingMessage;
-  readonly #channel: string | null;
-
-  constructor(message: Message.OutgoingMessage, channel: string | null) {
-    this.#message = message;
-    this.#channel = channel;
-  }
-
-  static createFromPromise(
-    messagePromise: Promise<Message.OutgoingMessage>,
-    channel: string | null
-  ): Promise<BiDiMessageEntry> {
-    return messagePromise.then(
-      (message) => new BiDiMessageEntry(message, channel)
+export class BidiServer extends EventEmitter<BidiServerEvents> {
+  #messageQueue: ProcessingQueue<OutgoingBidiMessage>;
+  #transport: ITransport;
+  #commandProcessor: CommandProcessor;
+  #eventManager: EventManager;
+  #cdpConnection: CdpConnection;
+  #selfTargetId: string;
+
+  constructor(
+    transport: ITransport,
+    cdpConnection: CdpConnection,
+    selfTargetId: string
+  ) {
+    super();
+    this.#selfTargetId = selfTargetId;
+    this.#cdpConnection = cdpConnection;
+    this.#messageQueue = new ProcessingQueue<OutgoingBidiMessage>(
+      this.#processMessageQueue
+    );
+    this.#transport = transport;
+    this.#transport.setOnMessage(this.#handleIncomingMessage);
+    this.#eventManager = new EventManager(this);
+    this.#commandProcessor = new CommandProcessor(
+      cdpConnection,
+      this.#eventManager,
+      this.#selfTargetId
+    );
+    this.#commandProcessor.on(
+      'response',
+      (response: Promise<OutgoingBidiMessage>) => {
+        this.emitOutgoingMessage(response);
+      }
     );
   }
 
-  static createResolved(
-    message: Message.OutgoingMessage,
-    channel: string | null
-  ): Promise<BiDiMessageEntry> {
-    return Promise.resolve(new BiDiMessageEntry(message, channel));
-  }
-
-  get message(): Message.OutgoingMessage {
-    return this.#message;
-  }
+  public async start() {
+    const cdpClient = this.#cdpConnection.browserClient();
 
-  get channel(): string | null {
-    return this.#channel;
-  }
-}
+    // Needed to get events about new targets.
+    await cdpClient.sendCommand('Target.setDiscoverTargets', {discover: true});
 
-export class BidiServer
-  extends EventEmitter<BidiServerEvents>
-  implements IBidiServer
-{
-  #messageQueue: ProcessingQueue<BiDiMessageEntry>;
+    // Needed to automatically attach to new targets.
+    await cdpClient.sendCommand('Target.setAutoAttach', {
+      autoAttach: true,
+      waitForDebuggerOnStart: true,
+      flatten: true,
+    });
 
-  constructor(private _transport: ITransport) {
-    super();
-    this.#messageQueue = new ProcessingQueue<BiDiMessageEntry>((m) =>
-      this.#sendMessage(m)
+    await Promise.all(
+      BrowsingContextStorage.getTopLevelContexts().map((c) => c.awaitLoaded())
     );
-    this._transport.setOnMessage(this.#onBidiMessage);
   }
 
-  async #sendMessage(messageEntry: BiDiMessageEntry) {
+  #processMessageQueue = async (messageEntry: OutgoingBidiMessage) => {
```

Comment:
```suggestion
  async #processMessageQueue(messageEntry: OutgoingBidiMessage) {
```

Reply:
The current style is needed so that explicit bind in the constructor is not needed.